### PR TITLE
Implement the gas estimator components refactor

### PR DESCRIPTION
**Pull request by silaslenihan**
Implementing to Gas Estimator Refactors as outlined in [this document](https://smartcontract-it.atlassian.net/wiki/spaces/DEPLOY/pages/573999065/Gas+Estimator+Refactor).

### CODE REVIEW

File: core/chains/evm/gas/arbitrum_estimator.go [hunk 1]
```diff
@@ -52,20 +43,26 @@ type arbitrumEstimator struct {
...
-func NewArbitrumEstimator(lggr logger.Logger, cfg ArbConfig, rpcClient rpcClient, ethClient ethClient) EvmEstimator {
+func NewArbitrumEstimator(lggr logger.Logger, cfg ArbConfig, ethClient ethClient) EvmEstimator {
 	lggr = logger.Named(lggr, "ArbitrumEstimator")
+
+	l1Oracle := rollups.NewArbitrumL1GasOracle(lggr, ethClient)
+
 	return &arbitrumEstimator{
 		cfg:            cfg,
-		EvmEstimator:   NewSuggestedPriceEstimator(lggr, rpcClient, cfg),
+		EvmEstimator:   NewSuggestedPriceEstimator(lggr, ethClient, cfg, "arbitrum"),
```
Comment (dimriou) [inline core/chains/evm/gas/arbitrum_estimator.go:57]: Change "arbitrum" to chaintype `ChainArbitrum` instead.

File: core/chains/evm/gas/fixed_price_estimator.go [hunk 2]
```diff
@@ -43,8 +47,12 @@ type fixedPriceEstimatorBlockHistoryConfig interface {
 
 // NewFixedPriceEstimator returns a new "FixedPrice" estimator which will
 // always use the config default values for gas prices and limits
-func NewFixedPriceEstimator(cfg fixedPriceEstimatorConfig, bhCfg fixedPriceEstimatorBlockHistoryConfig, lggr logger.Logger) EvmEstimator {
-	return &fixedPriceEstimator{cfg, bhCfg, logger.Sugared(logger.Named(lggr, "FixedPriceEstimator"))}
+func NewFixedPriceEstimator(cfg fixedPriceEstimatorConfig, ethClient evmclient.Client, bhCfg fixedPriceEstimatorBlockHistoryConfig, lggr logger.Logger, chainType config.ChainType) EvmEstimator {
+	var l1Oracle rollups.L1Oracle
+	if rollups.IsRollupWithL1Support(chainType) {
+		l1Oracle = rollups.NewL1GasOracle(lggr, ethClient, chainType)
```
Comment (dimriou) [inline core/chains/evm/gas/fixed_price_estimator.go:53]: Something that I don't quite like with the L1Oracle being created inside the Estimators is that it forces a client object to be passed. So in FixedPriceEstimator's case, you pass an `ethClient` that isn't actually used by the component, it' just needed for L1Oracle initialization. I'll review the rest, and come back with a final comment on this.

File: .changeset/flat-guests-marry.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"chainlink": minor
+---
+
+Gas Estimator L1GasOracle refactor
```
Comment (amit-momin) [inline .changeset/flat-guests-marry.md:5]: Could you also include that we're deprecating the arbgas cmd as a second point in here? Even though it hasn't really been used recently, might be a good to call out anyways

File: core/chains/evm/gas/models.go [hunk 4]
```diff
@@ -41,6 +43,13 @@ type EvmFeeEstimator interface {
 	GetMaxCost(ctx context.Context, amount assets.Eth, calldata []byte, feeLimit uint64, maxFeePrice *assets.Wei, opts ...feetypes.Opt) (*big.Int, error)
 }
 
+//go:generate mockery --quiet --name ethClient --output ./mocks/ --case=underscore --structname ETHClient
+type ethClient interface {
```
Comment (amit-momin) [inline core/chains/evm/gas/models.go:47]: To avoid some of the confusion we had before around the client, think it'd be helpful to rename this more specifically for the component. Maybe something like `feeEstimatorClient`?

File: core/chains/evm/gas/models.go [hunk 5]
```diff
@@ -41,6 +43,13 @@ type EvmFeeEstimator interface {
...
+	CallContract(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+	BatchCallContext(ctx context.Context, b []rpc.BatchElem) error
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
 // NewEstimator returns the estimator for a given config
 func NewEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, geCfg evmconfig.GasEstimator) EvmFeeEstimator {
```
Comment (amit-momin) [inline core/chains/evm/gas/models.go:54]: Then we should use that here so we're using the local interface
```suggestion
func NewEstimator(lggr logger.Logger, ethClient feeEstimatorClient, cfg Config, geCfg evmconfig.GasEstimator) EvmFeeEstimator {
```

File: core/chains/evm/gas/models.go [hunk 6]
```diff
@@ -138,6 +147,8 @@ type EvmEstimator interface {
 	//   - be sorted in order from highest price to lowest price
 	//   - all be of transaction type 0x2
 	BumpDynamicFee(ctx context.Context, original DynamicFee, maxGasPriceWei *assets.Wei, attempts []EvmPriorAttempt) (bumped DynamicFee, err error)
+
+	L1Oracle() rollups.L1Oracle
```
Comment (amit-momin) [inline core/chains/evm/gas/models.go:151]: Think we should return the pointer here. I probably miss this in the write up. Otherwise we could be dereferencing a nil l1Oracle in the estimator

File: core/chains/evm/gas/rollups/arbitrum_l1_oracle.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+	}
+	if gasCostErr != nil {
+		panic("Failed to parse L1 gas cost method ABI for chain: arbitrum")
+	}
+
+	return &arbitrumL1Oracle{
+		client:     ethClient,
+		pollPeriod: PollPeriod,
+		logger:     logger.Sugared(logger.Named(lggr, "L1GasOracle(arbitrum)")),
+		chainType:  "arbitrum",
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:103]: We should use the config.ChainType `ChainArbitrum` here. We maintain these chain types [here](https://github.com/smartcontractkit/chainlink/blob/b3aab74aed0dcc958277794d1ba94dc0d051dbfe/common/config/chaintype.go#L12-L23)

File: core/chains/evm/gas/rollups/arbitrum_l1_oracle.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+		l1GasPriceAddress:   l1GasPriceAddress,
+		gasPriceMethod:      gasPriceMethod,
+		l1GasPriceMethodAbi: l1GasPriceMethodAbi,
+		l1GasCostAddress:    l1GasCostAddress,
+		gasCostMethod:       gasCostMethod,
+		l1GasCostMethodAbi:  l1GasCostMethodAbi,
+
+		priceReader: nil,
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:112]: Since this is always nil, I think we can exclude priceReader logic in this l1Oracle. Think that was Optimism specific anyways.

File: core/chains/evm/gas/rollups/l1_oracle.go [hunk 9]
```diff
@@ -117,253 +43,19 @@ func IsRollupWithL1Support(chainType config.ChainType) bool {
...
 
-func NewL1GasOracle(lggr logger.Logger, ethClient ethClient, chainType config.ChainType) L1Oracle {
-	var priceReader daPriceReader
+func NewL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, chainType config.ChainType) L1Oracle {
+	var l1Oracle L1Oracle
 	switch chainType {
 	case config.ChainOptimismBedrock:
-		priceReader = newOPPriceReader(lggr, ethClient, chainType, OPGasOracleAddress)
+		l1Oracle = NewOpStackL1GasOracle(lggr, ethClient, chainType)
 	case config.ChainKroma:
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/l1_oracle.go:51]: Since this is the same as OP, could we just combine Kroma in the case above?

File: core/chains/evm/gas/rollups/l1_oracle_test.go [hunk 10]
```diff
@@ -76,7 +76,7 @@ func TestL1Oracle_GasPrice(t *testing.T) {
 		priceReader := mocks.NewDAPriceReader(t)
 		priceReader.On("GetDAGasPrice", mock.Anything).Return(l1BaseFee, nil)
 
-		oracle := newL1GasOracle(logger.Test(t), nil, config.ChainKroma, priceReader)
+		oracle := newOpStackL1GasOracle(logger.Test(t), nil, priceReader, "kroma")
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/l1_oracle_test.go:79]: We could probably also use the `config.ChainType` `ChainKroma` in tests too

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+	// `function getL1Fee(bytes) external view returns (uint256);`
+	OPGasOracle_getL1Fee = "getL1Fee"
+
+	// GasOracleAddress is the address of the precompiled contract that exists on Kroma chain.
+	// This is the case for Kroma.
+	KromaGasOracleAddress = "0x4200000000000000000000000000000000000005"
+	// GasOracle_l1BaseFee is the a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	KromaGasOracle_l1BaseFee = "l1BaseFee"
+)
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:102]: Think we have some duplicated consts here that we can delete.
```suggestion
const (
	// OPStackGasOracle_isEcotone fetches if the OP Stack GasPriceOracle contract has upgraded to Ecotone
	OPStackGasOracle_isEcotone = "isEcotone"
	// OPStackGasOracle_getL1GasUsed fetches the l1 gas used for given tx bytes
	OPStackGasOracle_getL1GasUsed = "getL1GasUsed"
	// OPStackGasOracle_isEcotonePollingPeriod is the interval to poll if chain has upgraded to Ecotone
	// Set to poll every 4 hours
	OPStackGasOracle_isEcotonePollingPeriod = 14400
	// OPStackGasOracleAddress is the address of the precompiled contract that exists on OP stack chain.
	// OPStackGasOracle_l1BaseFee fetches the l1 base fee set in the OP Stack GasPriceOracle contract
	// OPStackGasOracle_l1BaseFee is a hex encoded call to:
	// `function l1BaseFee() external view returns (uint256);`
	OPStackGasOracle_l1BaseFee = "l1BaseFee"
	// OPStackGasOracle_getL1Fee fetches the l1 fee for given tx bytes
	// OPStackGasOracle_getL1Fee is a hex encoded call to:
	// `function getL1Fee(bytes) external view returns (uint256);`
	OPStackGasOracle_getL1Fee = "getL1Fee"

        // This is the case for Optimism and Base.
	OPGasOracleAddress = "0x420000000000000000000000000000000000000F"
	// GasOracleAddress is the address of the precompiled contract that exists on Kroma chain.
	// This is the case for Kroma.
	KromaGasOracleAddress = "0x4200000000000000000000000000000000000005"
)
```

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+	priceReader := newOPPriceReader(lggr, ethClient, chainType, precompileAddress)
+	return newOpStackL1GasOracle(lggr, ethClient, priceReader, chainType)
+}
+
+func newOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, priceReader daPriceReader, chainType config.ChainType) L1Oracle {
+	var l1GasPriceAddress, gasPriceMethod, l1GasCostAddress, gasCostMethod string
+	var l1GasPriceMethodAbi, l1GasCostMethodAbi abi.ABI
+	var gasPriceErr, gasCostErr error
+
+	l1GasPriceAddress = OPGasOracleAddress
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:123]: Think we'll also want to use the `precompileAddress` from `NewOpStackL1GasOracle` in this method too so we call the right contract. Otherwise from Kroma we'd be calling the wrong address.

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+	logger     logger.SugaredLogger
+	chainType  config.ChainType
+
+	l1GasPriceAddress   string
+	gasPriceMethod      string
+	l1GasPriceMethodAbi abi.ABI
+	l1GasPriceMu        sync.RWMutex
+	l1GasPrice          priceEntry
+
+	l1GasCostAddress   string
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:43]: I think we can remove this field from this l1 oracle. This was introduced in the generic one to have a distinction between the contract we call for gas price vs gas cost because Arbitrum had separate contracts. Now that we have chain specific l1 oracles, we don't have to support that here for OP. We can also rename `l1GasPriceAddress` to something more generic like `l1OracleAddress` since all of the method are served out of the same contract. You'll notice we set the same address for both below in the builder anyways.

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+	isEcotoneMethodAbi abi.ABI
+
+	l1BaseFeeCalldata    []byte
+	isEcotoneCalldata    []byte
+	getL1GasUsedCalldata []byte
+	getL1FeeCalldata     []byte
+
+	isEcotone        bool
+	isEcotoneCheckTs int64
+}
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:68]: I think we can move all of the logic and fields from the `opStackGasPriceReader` to `optimismL1Oracle`. Think it'd improve readability if everything was consolidated into a single struct. The reason this was created as a separate struct before was because it was hard to incorporate in the generic l1 oracle. Now that we have an OP specific oracle we can just merge these. You'll need to consolidate the builder methods below into one too.

File: core/chains/evm/gas/rollups/op_price_reader_test.go
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_price_reader_test.go:1]: nit: Since we're merging the price reader with the l1 oracle, we should rename this to follow the usual standard of `op_l1_oracle_test.go`

File: core/chains/evm/gas/rollups/scroll_l1_oracle.go
Comment (amit-momin) [inline core/chains/evm/gas/rollups/scroll_l1_oracle.go:1]: Huh after reading through this l1 oracle, it looks like its the same as OP and Kroma too. We can maybe do the same thing we did with Kroma and just move this l1 oracle's precompile address into the OP stack one and just set the address using the chain type in the switch in the builder.

File: core/chains/evm/gas/suggested_price_estimator.go [hunk 15]
```diff
@@ -31,17 +33,12 @@ type suggestedPriceConfig interface {
...
-	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
-}
-
 // SuggestedPriceEstimator is an Estimator which uses the suggested gas price from eth_gasPrice.
 type SuggestedPriceEstimator struct {
 	services.StateMachine
 
 	cfg        suggestedPriceConfig
-	client     rpcClient
+	client     ethClient
```
Comment (amit-momin) [inline core/chains/evm/gas/suggested_price_estimator.go:41]: If I understood Dimitris suggestion for go best practices, maybe we want to keep a local client here but make it component specific like `suggestedPriceEstimatorClient`. Then we can use that in the struct and builder below so we really drill down that this component only needs `CallContext`.

File: core/chains/evm/gas/fixed_price_estimator.go (hunk 2 above)
Comment (amit-momin) [reply core/chains/evm/gas/fixed_price_estimator.go:53]: Not sure if we can get around that since we wanted l1Oracles to be embedded in estimators after this change but I understand the hesitation. Open to suggestions though if you have any after reading the rest of the PR.

File: core/chains/evm/gas/fixed_price_estimator.go [hunk 16]
```diff
@@ -43,8 +47,12 @@ type fixedPriceEstimatorBlockHistoryConfig interface {
 
 // NewFixedPriceEstimator returns a new "FixedPrice" estimator which will
 // always use the config default values for gas prices and limits
-func NewFixedPriceEstimator(cfg fixedPriceEstimatorConfig, bhCfg fixedPriceEstimatorBlockHistoryConfig, lggr logger.Logger) EvmEstimator {
-	return &fixedPriceEstimator{cfg, bhCfg, logger.Sugared(logger.Named(lggr, "FixedPriceEstimator"))}
+func NewFixedPriceEstimator(cfg fixedPriceEstimatorConfig, ethClient evmclient.Client, bhCfg fixedPriceEstimatorBlockHistoryConfig, lggr logger.Logger, chainType config.ChainType) EvmEstimator {
```
Comment (amit-momin) [inline core/chains/evm/gas/fixed_price_estimator.go:50]: We probably want to replace the `evmclient.Client` type here with the local interface

File: core/chains/evm/gas/fixed_price_estimator.go [hunk 17]
```diff
@@ -43,8 +47,12 @@ type fixedPriceEstimatorBlockHistoryConfig interface {
...
+	var l1Oracle rollups.L1Oracle
+	if rollups.IsRollupWithL1Support(chainType) {
```
Comment (amit-momin) [inline core/chains/evm/gas/fixed_price_estimator.go:52]: Curious what you think but what if we move this check into the `NewL1GasOracle` method so you don't have to have this check in all of the estimator builders? You can update `NewL1GasOracle` to return a pointer so it can return `nil` if this check doesn't pass.

File: core/chains/evm/gas/arbitrum_estimator.go [hunk 18]
```diff
@@ -110,7 +107,6 @@ func (a *arbitrumEstimator) GetLegacyGas(ctx context.Context, calldata []byte, l
 	if err != nil {
 		return
 	}
-	gasPrice = a.gasPriceWithBuffer(gasPrice, maxGasPriceWei)
```
Comment (amit-momin) [inline core/chains/evm/gas/arbitrum_estimator.go:113]: Since we reverted the Broadcaster changes, we may want to add `gasPriceWithBuffer` back to the arb estimator

File: core/chains/evm/gas/models.go (hunk 5 above)
Comment (silaslenihan) [reply core/chains/evm/gas/models.go:54]: I had to add `ConfiguredChainID()` and `HeadByNumber` to the `feeEstimatorClient` interface.

File: core/chains/evm/gas/rollups/op_l1_oracle.go (hunk 13 above)
Comment (silaslenihan) [reply core/chains/evm/gas/rollups/op_l1_oracle.go:43]: Ok, removing for Scroll too

**Comment by silaslenihan** [comment]
check if there are unused mocks in this package like [here](https://github.com/smartcontractkit/chainlink/tree/develop/core/chains/evm/gas/mocks) and [here](https://github.com/smartcontractkit/chainlink/tree/develop/core/chains/evm/gas/rollups/mocks)? Let’s make sure we cleanup legacy mocks if they aren’t used.

File: core/chains/evm/gas/models.go (hunk 6 above)
Comment (silaslenihan) [reply core/chains/evm/gas/models.go:152]: Note: We decided against this after further review

File: core/chains/evm/gas/suggested_price_estimator.go (hunk 15 above)
Comment (silaslenihan) [reply core/chains/evm/gas/suggested_price_estimator.go]: I added this back in as you suggested, but kept the client in the `NewSuggestedPriceEstimator` method signature parameters as `feeEstimatorClient` since that's what `NewL1GasOracle` requires.

File: core/chains/evm/gas/fixed_price_estimator.go (hunk 17 above)
Comment (silaslenihan) [reply core/chains/evm/gas/fixed_price_estimator.go:51]: Moved into `NewL1GasOralce`, however, we are no longer using the pointer as the return value. It still works though.

File: .changeset/flat-guests-marry.md [hunk 19]
```diff
@@ -0,0 +1,6 @@
+---
+"chainlink": minor
+---
+
+#changed Gas Estimator L1Oracles to be chain specific
```
Comment (amit-momin) [inline .changeset/flat-guests-marry.md:5]: First time seeing these tags but think we should just mark this as `#internal` since the functionality should still be the same

File: core/chains/evm/gas/models.go [hunk 20]
```diff
@@ -75,24 +85,24 @@ func NewEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, ge
 	switch s {
 	case "Arbitrum":
 		newEstimator = func(l logger.Logger) EvmEstimator {
-			return NewArbitrumEstimator(lggr, geCfg, ethClient, ethClient)
+			return NewArbitrumEstimator(lggr, geCfg, ethClient)
 		}
 	case "BlockHistory":
 		newEstimator = func(l logger.Logger) EvmEstimator {
 			return NewBlockHistoryEstimator(lggr, ethClient, cfg, geCfg, bh, *ethClient.ConfiguredChainID())
```
Comment (amit-momin) [inline core/chains/evm/gas/models.go:92]: Noticed we had to add `ConfiguredChainID` to the interface just for this builder. I think the `ChainID` should be available upstream in the `Config` object. I don't want scope creep to hold this PR up anymore so I think it's ok the way it is. Just adding this note here so anyone else reviewing knows we considered it.

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 21]
```diff
@@ -0,0 +1,440 @@
...
+
+	"github.com/smartcontractkit/chainlink/v2/common/client"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+)
+
+type OptimismL1Oracle interface {
+	L1Oracle
+	GetDAGasPrice(ctx context.Context) (*big.Int, error)
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:31]: I think we only need this method internally so we don't have to expose this here. Which also means we don't really need this custom interface either like Arb.

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 22]
```diff
@@ -0,0 +1,440 @@
...
+	// This is the case for Optimism and Base.
+	OPGasOracleAddress = "0x420000000000000000000000000000000000000F"
+	// GasOracleAddress is the address of the precompiled contract that exists on Kroma chain.
+	// This is the case for Kroma.
+	KromaGasOracleAddress = "0x4200000000000000000000000000000000000005"
+	// ScrollGasOracleAddress is the address of the precompiled contract that exists on Scroll chain.
+	ScrollGasOracleAddress = "0x5300000000000000000000000000000000000002"
+)
+
+func NewOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, chainType config.ChainType) OptimismL1Oracle {
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:92]: To go with the previous comment, that means we can just directly return:
```suggestion
func NewOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, chainType config.ChainType) L1Oracle {
```

File: core/chains/evm/gas/rollups/op_l1_oracle.go [hunk 23]
```diff
@@ -0,0 +1,440 @@
...
+	}
+
+	o.l1GasPriceMu.Lock()
+	defer o.l1GasPriceMu.Unlock()
+	o.l1GasPrice = priceEntry{price: assets.NewWei(price), timestamp: time.Now()}
+	return
+}
+
+func (o *optimismL1Oracle) fetchL1GasPrice(ctx context.Context) (price *big.Int, err error) {
+	return o.GetDAGasPrice(ctx)
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/op_l1_oracle.go:254]: Think we can just copy the stuff inside of `GetDAGasPrice` here since this is all it does now.

File: core/chains/evm/gas/rollups/op_l1_oracle.go (hunk 21 above)
Comment (silaslenihan) [reply core/chains/evm/gas/rollups/op_l1_oracle.go:31]: Only reason I did this is because in `op_l1_oracle_test.go` `GetDAGasPrice` is directly referenced

File: core/chains/evm/gas/rollups/op_l1_oracle.go (hunk 23 above)
Comment (silaslenihan) [reply core/chains/evm/gas/rollups/op_l1_oracle.go:254]: As per my reply above, this would effect `op_l1_oracle_test.go`

File: core/chains/evm/gas/rollups/op_l1_oracle.go (hunk 21 above)
Comment (amit-momin) [reply core/chains/evm/gas/rollups/op_l1_oracle.go:31]: Ah gotcha, ok we can keep it like this then

File: core/chains/evm/gas/rollups/op_l1_oracle.go (hunk 23 above)
Comment (amit-momin) [reply core/chains/evm/gas/rollups/op_l1_oracle.go:254]: True. Ok in that case we can just call `GetDAGasPrice` directly wherever `fetchL1GasPrice` was being called.

File: core/chains/evm/gas/rollups/op_l1_oracle.go (hunk 21 above)
Comment (silaslenihan) [reply core/chains/evm/gas/rollups/op_l1_oracle.go:31]: Sounds good

**Review by amit-momin — DISMISSED** [review]
LGTM!

File: core/chains/evm/gas/arbitrum_estimator.go [hunk 24]
```diff
@@ -28,19 +25,14 @@ type ArbConfig interface {
...
-
 // arbitrumEstimator is an Estimator which extends SuggestedPriceEstimator to use getPricesInArbGas() for gas limit estimation.
 type arbitrumEstimator struct {
 	services.StateMachine
 	cfg ArbConfig
 
 	EvmEstimator // *SuggestedPriceEstimator
 
-	client     ethClient
+	client     feeEstimatorClient
```
Comment (dimriou) [inline core/chains/evm/gas/arbitrum_estimator.go:35]: No need to store `client` in a variable. It's being passed inside the suggested estimator and the l1oracle.

File: core/chains/evm/gas/models.go [hunk 25]
```diff
@@ -75,24 +85,24 @@ func NewEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, ge
...
+			return NewSuggestedPriceEstimator(lggr, ethClient, geCfg, cfg.ChainType())
 		}
 	default:
 		lggr.Warnf("GasEstimator: unrecognised mode '%s', falling back to FixedPriceEstimator", s)
 		newEstimator = func(l logger.Logger) EvmEstimator {
-			return NewFixedPriceEstimator(geCfg, bh, lggr)
+			return NewFixedPriceEstimator(geCfg, ethClient, bh, lggr, cfg.ChainType())
 		}
 	}
 	return NewWrappedEvmEstimator(lggr, newEstimator, df, l1Oracle, geCfg)
```
Comment (dimriou) [inline core/chains/evm/gas/models.go:108]: Might be out of scope, but **why** do we pass a function that creates an estimator instead of the actual estimator 🤔 ?

Comment (dimriou) [reply core/chains/evm/gas/models.go:108]: Side note: we have `EvmFeeEstimator`,  `EvmEstimator`, `WrappedEvmEstimator` which is quite confusing. `WrappedEvmEstimator` seems like the implementation of `EvmFeeEstimator` interface and its scope is local. Probably we can rename it to `evmFeeEstimator` and make it private, so by looking you can quickly understand the association.

**Review by dimriou — COMMENTED** [review]
`L1Oracle` interface should live inside `l1_oracle.go`, not in a separate models.go file. Let's remove the interface there. and delete the unnecessary file.

File: core/chains/evm/gas/models.go
Comment (dimriou) [inline core/chains/evm/gas/models.go:1]: We do duplicate work when creating L1Oracles. First we call this: 
`if rollups.IsRollupWithL1Support(cfg.ChainType()) {
		l1Oracle = rollups.NewL1GasOracle(lggr, ethClient, cfg.ChainType())
	}`
and then inside the estimators we do the same thing. Instead we should be passing the L1Oracle object, otherwise this creates two different instances of the service. This will also simplify the client interface that we need to pass inside the estimators.

Comment (dimriou) [reply core/chains/evm/gas/models.go:1]: Also this code probably doesn't work, because the l1oracle created inside the estimators is never started.

File: core/chains/evm/gas/models.go (hunk 25 above)
Comment (silaslenihan) [reply core/chains/evm/gas/models.go:108]: > but why do we pass a function that creates an estimator instead of the actual estimator 🤔
I'm not sure, but this is preexisting functionality so I believe it's out of scope for this PR

Comment (silaslenihan) [reply core/chains/evm/gas/models.go:108]: > Probably we can rename it to evmFeeEstimator and make it private, so by looking you can quickly understand the association.

Agreed, will implement this change

File: core/chains/evm/gas/models.go
Comment (silaslenihan) [reply core/chains/evm/gas/models.go:1]: Yes, you're right I had duplicate code of creating the `L1Oracles`. I think instead, we should just be creating an l1Oracle within each of the estimators. The wrapped Evm estimator need not store this `L1Oracle`

Comment (silaslenihan) [reply core/chains/evm/gas/models.go:1]: The `L1Oracle` should be started [here](https://github.com/smartcontractkit/chainlink/blob/4a55a9378704e8600669f3d2182db59698bacc6a/core/chains/evm/gas/models.go#L199-L204) where the `evmFeeEstimator` (formerly `WrappedEvmEstimator`) is started. Would this not work?

Comment (silaslenihan) [reply core/chains/evm/gas/models.go:1]: Ah I see why this would cause a bug. Since we're returning an L1Oracle value [here](https://github.com/smartcontractkit/chainlink/blob/4a55a9378704e8600669f3d2182db59698bacc6a/core/chains/evm/gas/models.go#L252-L254) rather than a pointer, it will be a copy of the L1Oracle

File: core/chains/evm/gas/rollups/arbitrum_l1_oracle.go [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+	ArbNodeInterfaceAddress = "0x00000000000000000000000000000000000000C8"
+	// ArbGasInfo_getPricesInArbGas is the a hex encoded call to:
+	// `function gasEstimateL1Component(address to, bool contractCreation, bytes calldata data) external payable returns (uint64 gasEstimateForL1, uint256 baseFee, uint256 l1BaseFeeEstimate);`
+	ArbNodeInterface_gasEstimateL1Component = "gasEstimateL1Component"
+	// ArbGasInfo_getPricesInArbGas is the a hex encoded call to:
+	// `function getPricesInArbGas() external view returns (uint256, uint256, uint256);`
+	ArbGasInfo_getPricesInArbGas = "02199f34"
+)
+
+func NewArbitrumL1GasOracle(lggr logger.Logger, ethClient l1OracleClient) ArbL1GasOracle {
```
Comment (dimriou) [inline core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:74]: Why two constructors?

Comment (silaslenihan) [reply core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:74]: Removed this

File: core/chains/evm/gas/rollups/arbitrum_l1_oracle.go [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:74]: Following the pattern from the OP L1 oracle, should we return `*arbitrumL1Oracle` here instead?
```suggestion
func NewArbitrumL1GasOracle(lggr logger.Logger, ethClient l1OracleClient) *arbitrumL1Oracle {
```

Comment (silaslenihan) [reply core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:74]: Yes, I believe this is best practice

**Review by amit-momin — DISMISSED** [review]

File: core/chains/evm/gas/arbitrum_estimator.go [hunk 28]
```diff
@@ -52,20 +43,25 @@ type arbitrumEstimator struct {
...
+func NewArbitrumEstimator(lggr logger.Logger, cfg ArbConfig, ethClient feeEstimatorClient) EvmEstimator {
 	lggr = logger.Named(lggr, "ArbitrumEstimator")
+
+	l1Oracle := rollups.NewArbitrumL1GasOracle(lggr, ethClient)
+
 	return &arbitrumEstimator{
 		cfg:            cfg,
-		EvmEstimator:   NewSuggestedPriceEstimator(lggr, rpcClient, cfg),
-		client:         ethClient,
+		EvmEstimator:   NewSuggestedPriceEstimator(lggr, ethClient, cfg, config.ChainArbitrum),
```
Comment (dimriou) [inline core/chains/evm/gas/arbitrum_estimator.go:57]: One problem I see with creating the l1Oracles inside the estimators instead of injecting them is that in this case you will create a new L1Oracle inside the suggested price estimator and eventually you will have two arbitrum oracles making RPC requests. It probably won't affect the functionality, but you do end up having two services in the background doing the exact same thing.

File: core/chains/evm/gas/rollups/arbitrum_l1_oracle.go [hunk 29]
```diff
@@ -0,0 +1,301 @@
...
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/v2/common/client"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+)
+
+//go:generate mockery --quiet --name ArbL1GasOracle --output ./mocks/ --case=underscore --structname ArbL1GasOracle
```
Comment (dimriou) [inline core/chains/evm/gas/rollups/arbitrum_l1_oracle.go:28]: Remove this, it's unused.

File: core/chains/evm/gas/rollups/mocks/arb_l1_gas_oracle.go
Comment (dimriou) [inline core/chains/evm/gas/rollups/mocks/arb_l1_gas_oracle.go:1]: Delete this mock, it's unused.

**Review by dimriou — APPROVED** [review]

**Review by amit-momin — APPROVED** [review]
